test/unit/spread.spec.ts

```diff
@@ -33,6 +48,32 @@ describe("in array literal", () => {
         [tstl.LuaTarget.Lua53]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
     });
 
+    test("of tuple return call", () => {
```

Comment:
Would be nice to test it in function call as well.
Maybe the duplication there could be reduced? With something like
```ts
describe.each(["function call",	"array literal"])("in %s", kind => {
	const factory = code => kind === "function call"
		? `((...args: any[]) => args)(${code})`
		: `[${code}]`;

	test.each(spreadCases)("of array literal (%p)", expression => {
        util.testExpression(factory(expression)).expectToMatchJsResult();
    });
});
```

Reply:
wew, I would rather have some duplication if it keeps our tests clear

Reply:
That's not just some duplication: since both cases follow a single code path pretty much all cases would be the same, except for `OmittedExpression` since it's valid only in arrays and target-specific tests. I don't really see how a simple factory function makes these cases harder to read, but it makes it easier to keep them consistent (for example even now `of tuple return` test has different name and function call doesn't have a `of iterable` case). That's how all of the current tests would look:

```ts
describe.each(["function call",	"array literal"])("in %s", kind => {
    const factory = code => kind === "function call"
        ? `((...args: any[]) => args)(${code})`
        : `[${code}]`;

    test.each(arrayLiteralCases)("of array literal (%p)", expression => {
        util.testExpression(factory(expression)).expectToMatchJsResult();
    });

    test.each(arrayLiteralCases)("of tuple return call", expression => {
        util.testFunction`
            ${tupleReturnDefinition}
            return ${factory(`...tuple(${expression})`)};
        `.expectToMatchJsResult();
    });

    test("of string literal", () => {
        util.testExpression(factory('..."spread", ..."string"')).expectToMatchJsResult();
    });

    // TODO:
    test.each(["", "string", "string with spaces", "string 1 2 3"])("of string literal (%p)", str => {
        util.testExpression(factory(`...${str}`)).expectToMatchJsResult();
    });

    test("of iterable", () => {
        util.testFunction`
            const it = {
                i: -1,
                [Symbol.iterator]() {
                    return this;
                },
                next() {
                    ++this.i;
                    return {
                        value: 2 ** this.i,
                        done: this.i == 9,
                    }
                }
            };

            return ${factory('...it')};
        `.expectToMatchJsResult();
    });
});

describe("in function call", () => {
    util.testEachVersion(
        undefined,
        () => util.testFunction`
            function foo(a: number, b: number, ...rest: number[]) {
                return { a, b, rest }
            }
            const array = [0, 1, 2, 3] as const;
            return foo(...array);
        `,
        {
            [tstl.LuaTarget.LuaJIT]: builder => builder.tap(expectUnpack),
            [tstl.LuaTarget.Lua51]: builder => builder.tap(expectUnpack),
            [tstl.LuaTarget.Lua52]: builder => builder.tap(expectTableUnpack),
            [tstl.LuaTarget.Lua53]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
        }
    );

});

describe("in array literal", () => {
    util.testEachVersion(undefined, () => util.testExpression`[...[0, 1, 2]]`, {
        [tstl.LuaTarget.LuaJIT]: builder => builder.tap(expectUnpack),
        [tstl.LuaTarget.Lua51]: builder => builder.tap(expectUnpack),
        [tstl.LuaTarget.Lua52]: builder => builder.tap(expectTableUnpack),
        [tstl.LuaTarget.Lua53]: builder => builder.tap(expectTableUnpack).expectToMatchJsResult(),
    });

    test("of array literal /w OmittedExpression", () => {
        util.testFunction`
            const array = [1, 2, ...[3], , 5];
            return { a: array[0], b: array[1], c: array[2], d: array[3] };
        `.expectToMatchJsResult();
    });
});
```